pcap/search.go

```diff
@@ -136,22 +136,63 @@ func genICMPFilter(src, dst net.IP) PacketFilter {
 	}
 }
 
-// XXX currently assumes legacy pcap is produced by the input reader
 // XXX need to handle searching over multiple pcap files
 func (s *Search) Run(ctx context.Context, w io.Writer, r pcapio.Reader) error {
+	reader := s.Reader(r)
+	_, err := reader.WriteToWithContext(ctx, w)
+	return err
+}
+
+type SearchReader struct {
+	*Search
+	reader pcapio.Reader
+	opts   gopacket.DecodeOptions
+	npkt   int
+}
+
+func (s *Search) Reader(r pcapio.Reader) *SearchReader {
 	opts := gopacket.DecodeOptions{Lazy: true, NoCopy: true}
-	var npkt int
+	return &SearchReader{Search: s, reader: r, opts: opts}
+}
+
+func (s *SearchReader) WriteToWithContext(ctx context.Context, w io.Writer) (n int64, err error) {
```

Comment:
The naming here doesn't align anymore: SearchReader should become "SearchFilter" or "SearchWriter" to match its Write-like functionality.

You could also pass in the context when a SearchFilter (or whatever) is created, then change this method to match the usual Writer interface.

Reply:
It actually is more like a `Reader` as it attempts to follow the `WriterTo` interface (see the godocs for `io.Copy`: https://golang.org/pkg/io/#Copy). I would've liked to make SearchReader an `io.Reader` but pkg `io` doesn't have an `io.CopyWithContext` func.

Reply:
I'm not sure this is an improvement, but you could make this WriterTo and use io.Copy if you wrap the writer with something like this:
```go
type ContextWriter struct {
	ctx context.Context
	w   io.Writer
}

func NewContextWriter(ctx context.Context, w io.Writer) *ContextWriter {
	return &ContextWriter{ctx: ctx, w: w}
}

func (c *ContextWriter) Write(buf []byte) (int, error) {
	if err := c.ctx.Err(); err != nil {
		return 0, err
	}
	return c.w.Write(buf)
}
```

Reply:
@nwt also an interesting approach.